### PR TITLE
feat(core): Create StaticInjector which does not depend on Reflect polyfill

### PR DESCRIPTION
**Pull request by mhevery**

### CODE REVIEW

File: packages/core/src/di/injector.ts [hunk 1]
```diff
@@ -60,4 +63,333 @@ export abstract class Injector {
...
+    this.parent = parent;
+    const records = this._records = <Records>{};
+    const id = getId(Injector);
+    records[id] = <Record>{token: Injector, fn: IDENT, deps: EMPTY, value: this, useNew: false};
+    recursivelyProcessProviders(records, providers);
+  }
+
+  get<T>(token: Type<T>|InjectionToken<T>, notFoundValue?: T): T;
+  get(token: any, notFoundValue?: any): any;
+  get(token: any, notFoundValue?: any): any {
```
Comment (devoto13) [inline packages/core/src/di/injector.ts]: Is duplicate signature intended?

Comment (mhevery) [inline packages/core/src/di/injector.ts]: Yes, see https://www.typescriptlang.org/docs/handbook/functions.html#overloads

File: packages/core/src/di/reflective_provider_interfaces.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * @whatItDoes Describes how the {@link Injector} should be configured.
+ * @howToUse
+ * See {@link TypeProvider}, {@link ClassProvider}, {@link StaticProvider}.
+ *
+ * @description
+ * For more details, see the {@linkDocs guide/dependency-injection "Dependency Injection Guide"}.
+ *
+ * @stable
+ */
+export type Provider =
```
Comment (tbosch) [inline packages/core/src/di/reflective_provider_interfaces.ts]: `Provider` is also used in our decorators (e.g. `@NgModule`). It seems wrong that it is now defined in a file called `reflective_injector_interfaces`. I think it should be defined in e.g. `di/metadata.ts` and then this file can reexport it if we really need this file.

File: packages/core/src/di/injector.ts [hunk 3]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+
+type TokenPath = Array<any>;
+
+function resolveProvider(provider: SupportedProvider): Record {
+  const deps = computeDeps(provider);
+  let fn: Function = IDENT;
+  let value: any = EMPTY;
+  let useNew: boolean = false;
+  let provide = resolveForwardRef(provider.provide);
+  if (USE_VALUE in provider) {
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Why use `in` here? For all the cases we can just check for e.g. `.useFactory !== undefined`, and the fall through case is `useValue`...

Comment (tbosch) [inline packages/core/src/di/injector.ts]: This would both be faster and remove `getClosureSafeProperty`

File: packages/core/src/di/injector.ts [hunk 4]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+}
+
+interface DependencyRecord {
+  id: string;
+  token: any;
+  options: number;
+}
+
+interface Records {
+  [key: string]: Record;
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Why use an object hash and not a `Map` directly?

File: packages/core/src/di/injector.ts [hunk 5]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+  } else {
+    value = parent.get(token, notFoundValue);
+  }
+  return value;
+}
+
+
+let idCounter = 0;
+
+function getId(obj: any): string {
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: I think using a `Map` is simpler, and shouldn't be slower AFAIK.

File: packages/core/src/di/injector.ts [hunk 6]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+              // Other records we know about.
+              records,
+              // If we don't know how to resolve dependency and we should not check parent for it,
+              // than pass in Null injector.
+              !childRecord && !(options & OPT_CHECK_PARENT) ? NULL_INJECTOR : parent,
+              options & OPT_OPTIONAL ? null : Injector.THROW_IF_NOT_FOUND));
+        }
+      }
+      if (useNew) {
+        obj = Object.create(fn.prototype);
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: This won't work if users run in real ES6 mode.
Consider doing a `obj = new fn(...deps)`.

File: packages/core/src/di/injector.ts [hunk 7]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+const USE_EXISTING =
+    getClosureSafeProperty<ExistingProvider>({provide: String, useExisting: GET_PROPERTY_NAME});
+const PROVIDE =
+    getClosureSafeProperty<StaticProvider>({provide: GET_PROPERTY_NAME, useValue: null});
+const MULTI = getClosureSafeProperty<ConstructorProvider>(
+    {provide: String, multi: GET_PROPERTY_NAME, deps: EMPTY});
+const NG_TOKEN_PATH = 'ngTokenPath';
+const NG_TEMP_TOKEN_PATH = 'ngTempTokenPath';
+const ID_EXPANDO = '__symbol_angular_id_' + new Date().getTime() + '__';
+const OPT_OPTIONAL = 1;
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Consider using an `enum` for the bitmask here and add the suffix `Flags` (which is what we do everywhere else):
```
enum OptionFlags {
  Optional = 1 << 0,
  CheckSelf = 1 << 1,
  CheckParent = 1 << 2,
}
```

File: packages/core/src/di/injector.ts [hunk 8]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+    token: any, record: Record | undefined, records: Records, parent: Injector,
+    notFoundValue: any): any {
+  try {
+    return resolveToken(token, record, records, parent, notFoundValue);
+  } catch (e) {
+    // ensure that 'e' is of type Error.
+    if (!(e instanceof Error)) {
+      e = new Error(e);
+    }
+    const path: any[] = e[NG_TEMP_TOKEN_PATH] = e[NG_TEMP_TOKEN_PATH] || [];
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Consider passing the `tokenPathOnError` directly to `tryResolveToken`, so you don't need to patch it unto the error (as it is a private field anyways).

File: packages/core/src/di/injector.ts [hunk 9]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+    if (value == CIRCULAR) {
+      throw Error(NO_NEW_LINE + 'Circular dependency');
+    } else if (value === EMPTY) {
+      record.value = CIRCULAR;
+      let obj = undefined, useNew = record.useNew, fn = record.fn, depRecords = record.deps,
+          deps = EMPTY;
+      if (depRecords.length) {
+        deps = [];
+        for (let i = 0; i < depRecords.length; i++) {
+          const depRecord: DependencyRecord = depRecords[i], options = depRecord.options,
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Maybe have separate `const` statements to clear up the formatting?

File: packages/core/src/di/injector.ts [hunk 10]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+  get<T>(token: Type<T>|InjectionToken<T>, notFoundValue?: T): T;
+  get(token: any, notFoundValue?: any): any;
+  get(token: any, notFoundValue?: any): any {
+    const record = this._records[getId(token)];
+    try {
+      return tryResolveToken(token, record, this._records, this.parent, notFoundValue);
+    } catch (e) {
+      const tokenPath: any[] = e[NG_TEMP_TOKEN_PATH];
+      let message: string = e.message;
+      message = message && message.charAt(0) === NO_NEW_LINE ? message.substr(1) : '\n' + message;
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Consider adding the logic of adding a newLine or not to `staticError`

File: packages/core/src/di/injector.ts [hunk 11]
```diff
@@ -60,4 +63,347 @@ export abstract class Injector {
...
+      const error = staticError(message, tokenPath);
+      e.message = error.message;
```
Comment (tbosch) [inline packages/core/src/di/injector.ts]: Did you check if this prints nicely in the browser console? AFAIK it does not.
Another solution would be to throw 2 errors: 1 for flow control where it is now, and another one here with the proper message passed to `new Error(...)`

File: packages/examples/core/di/ts/provider_spec.ts [hunk 12]
```diff
@@ -89,6 +89,61 @@ export function main() {
       });
     });
 
+    describe('StaticClassProvider', () => {
+      it('works', () => {
+        // #docregion StaticClassProvider
```
Comment (tbosch) [inline packages/examples/core/di/ts/provider_spec.ts]: I think this is missing a guide that refers to these new `#docregion`s...

**Review by tbosch — CHANGES_REQUESTED** [review]
Looks good in general, needs cleanup.

File: packages/core/src/di/injector.ts (hunk 3 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: I can't use `.useValue` since the value could be `undefined`. So since I started using `in` with `useValue`, I am just being consistent.

File: packages/core/src/di/injector.ts (hunk 4 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: See: https://stackoverflow.com/questions/10258071/class-object-vs-hashmap

Summary: faster, smaller, no polyfill

File: packages/core/src/di/injector.ts (hunk 11 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: Yes, it looks good in console.

File: packages/core/src/di/injector.ts (hunk 7 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: Done.

File: packages/core/src/di/injector.ts (hunk 10 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: Done

File: packages/core/src/di/injector.ts (hunk 8 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: I don't want to have the cost of `tokenPath` only in the case of an error. Hence it is patched on `Error`

File: packages/core/src/di/injector.ts (hunk 9 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: done

File: packages/core/src/di/injector.ts (hunk 6 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: good catch, fixed

File: packages/examples/core/di/ts/provider_spec.ts (hunk 12 above)
Comment (mhevery) [inline packages/examples/core/di/ts/provider_spec.ts]: Not sure I understand. It is present.

File: packages/core/src/di/injector.ts (hunk 5 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: changed

File: packages/core/src/di/injector.ts (hunk 4 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: changed to map

File: packages/core/src/di/injector.ts (hunk 3 above)
Comment (mhevery) [inline packages/core/src/di/injector.ts]: changed, only `USE_VALUE` is special

File: packages/core/src/di/reflective_provider_interfaces.ts (hunk 2 above)
Comment (mhevery) [inline packages/core/src/di/reflective_provider_interfaces.ts]: moved back to `provider.ts`

File: packages/core/src/di/injector.ts (hunk 8 above)
Comment (tbosch) [inline packages/core/src/di/injector.ts]: I mean you can just create it as an empty array, and only fill it in case of an error in the various `catch` clauses. This way, the cost is only to create the empty array, but not to fill it.

File: packages/examples/core/di/ts/provider_spec.ts (hunk 12 above)
Comment (tbosch) [inline packages/examples/core/di/ts/provider_spec.ts]: Ah, never mind, I remembered incorrectly how the guides work.

File: packages/benchpress/index.ts [hunk 13]
```diff
@@ -9,7 +9,7 @@
 // Must be imported first, because Angular decorators throw on load.
 import 'reflect-metadata';
 
-export {InjectionToken, Injector, Provider, ReflectiveInjector} from '@angular/core';
+export {InjectionToken, Injector, Provider, ReflectiveInjector, StaticProvider} from '@angular/core';
```
Comment (tbosch) [inline packages/benchpress/index.ts]: Why the changes in benchpress? We might have some code in G3 that relies on this supporting reflective injector...

Comment (tbosch) [inline packages/benchpress/index.ts]: Why the changes in benchpress? We might have some code in G3 that relies on this supporting reflective injector...

**Comment by tbosch** [comment]
Type in commit message of the second commit:
```
Specifically the method signature _when_ from `Provider[]` to ...
```

I think the second commit message deserves an example for migration, e.g. a `before` / `after` section.

**Comment by tbosch** [comment]
Also for the commit message of the second commit: The providers for the jit compiler, that users were able to also specify, are now `StaticProvider`s.

**Review by tbosch — APPROVED** [review]

File: packages/benchpress/index.ts (hunk 13 above)
Comment (tbosch) [inline packages/benchpress/index.ts]: Never mind, I think it is actually good to change benchpress as well.

**Comment by vicb** [comment]
blocked on tap/164481065

**Comment by mhevery** [comment]
@vicb here is the fix http://cl/164500104

**Review by IgorMinar — APPROVED** [review]

**Comment by trotyl** [comment]
Would https://github.com/angular/angular/issues/13820 be affected by this? Seems StaticInjector does not require `@Injectable()` anymore.

**Comment by sod** [comment]
`@Injectable()` is just a trick to get typescript to export the typings from the constructor, thanks to `emitDecoratorMetadata` setting in tsconfig.json. It'll most likley stay.

image

**Comment by trotyl** [comment]
@sod https://github.com/angular/angular/issues/13820 is about requiring `@Injectable()` when there is no dependency, where metadata is not needed here.

**Comment by sod** [comment]
Commit message https://github.com/angular/angular/commit/fcadbf4bf6d00ea5b250a8069e05b3e4bd000a29 reads like you are indeed right. They deprecated or even got rid of the reflective resolver altogether.
